src/base/random.cc

```diff
@@ -50,52 +50,45 @@ namespace gem5
 
 Random::Random()
 {
-    // default random seed
-    init(5489);
+    init(globalSeed);
+    instances.push_back(this);
 }
 
 Random::Random(uint32_t s)
 {
     init(s);
+    instances.push_back(this);
 }
 
-Random::~Random()
+Random::Random(const Random& rng)
 {
+    instances.push_back(this);
+    this->gen = rng.gen;
 }
 
-void
-Random::init(uint32_t s)
+Random& Random::operator=(const Random& rng)
 {
-    gen.seed(s);
+    instances.push_back(this);
+    this->gen = rng.gen;
+    return *this;
 }
 
-void
-Random::serialize(CheckpointOut &cp) const
+Random::~Random()
 {
-    panic("Currently not used anywhere.\n");
+    assert(std::find(instances.begin(), instances.end(), this) !=
+        instances.end());
 
-    // get the state from the generator
-    std::ostringstream oss;
-    oss << gen;
-    std::string state = oss.str();
-    paramOut(cp, "mt_state", state);
+    instances.erase(std::remove(
+        instances.begin(), instances.end(), this));
 }
 
 void
-Random::unserialize(CheckpointIn &cp)
+Random::init(uint32_t s)
 {
-    panic("Currently not used anywhere.\n");
-
-    // the random generator state did not use to be part of the
-    // checkpoint state, so be forgiving in the unserialization and
-    // keep on going if the parameter is not there
-    std::string state;
-    if (optParamIn(cp, "mt_state", state)) {
-        std::istringstream iss(state);
-        iss >> gen;
-    }
+    gen.seed(s);
 }
 
-Random random_mt;
+uint64_t Random::globalSeed = 5489;
+std::vector<Random*> Random::instances = {};
```

Comment:
IMHO we shouldn't be relying on unique_ptrs. They are used to model exclusive ownership of the pointer holder towards the pointed data. Here the static vector does not own the random generators which are instead owned by the objects instantiating them. The vector just stores a non-owning pointer so that it can pass them the new seed once it gets generated,

Also as @aperais pointed out, we cannot make any assumption about the lifetime of the Random object.
You can literally instantiate the Random object on the stack if you wanted to and to destroy it automatically when it goes out of scope. This is not compatible with unique_ptr lifetime management

Reply:
Thanks for the info. This makes sense to me.

I do wonder if it is a bad code smell to "instantiate the Random object on the stack if you wanted to and to destroy it automatically when it goes out of scope." It seems that it may be a good idea to require things that use randomness in gem5 to follow a more restricted model. But I'm happy to say that's out of scope.

Reply:
> Thanks for the info. This makes sense to me.
> 
> I do wonder if it is a bad code smell to "instantiate the Random object on the stack if you wanted to and to destroy it automatically when it goes out of scope." It seems that it may be a good idea to require things that use randomness in gem5 to follow a more restricted model. But I'm happy to say that's out of scope.

Agreed and it was just an hypothetical example. @aperais gives a better one which is actually how the random generator is mainly used throughout the code

> class A
> {
>   Random rng = {};
> }
> 
> So the storage for the Random object is allocated when instance of A is created and freed when instance of A is deleted. If you store a unique_ptr<Random> in the instances vector, I guess you are opening yourself to a double free because the storage for Random will be freed when instances is deleted (at exit I suppose) and when the instance of A gets deleted